README.md
# XCLANPro - Under Developing
![XCLANPro_home2](https://user-images.githubusercontent.com/67137156/170579752-e92115d1-9c53-457b-93ea-1539a5d36044.png)
![XCLANPro_conf](https://user-images.githubusercontent.com/67137156/170580003-3cc3b607-fe0f-4f3c-bf86-a11b71956def.png)

## info about the script :point_right: [XCLANPro](https://github.com/Farukbrowser/XCLANPro/blob/main/info.md)

## Telegram key generator bot :point_right: [@XCLANPro_bot](https://t.me/XCLANPro_bot)

## Install XCLANPro

`rm -rf install.sh* && wget https://raw.githubusercontent.com/Farukbrowser/XCLANPro/main/install.sh && chmod 775 install.sh* && ./install.sh* --start`

## Update XCLANPro

`rm -rf install.sh* && wget https://raw.githubusercontent.com/Farukbrowser/XCLANPro/main/install.sh && chmod +x install.sh* && ./install.sh* --update`

## UPDATE

2023-06-03
4)- fix users without internet (beta)

 2023-05-10
1)- added psiphon protocol
 2023-04-25
 
1)- Added udp-custom protocol, apps(HTTPCustom)

 2023-04-07
 
1)- UDPserver, ip selection option nat and public ip detection
2023-03-31
 
1)- fix user SSH/TOKEN/HWID, correction when renewing does not give data.

2023-01-06

1)- fix user SSH/TOKEN/HWID, visualizacion en orde alfabetico

2)- fix bugs renovar usuarios unlock.

2022-12-24

1)- Nuevo script tcpbbr, nucles mas actuales, instalacion mas segura

2022-12-19

1)- fix code all script ssl

2)- se inabilito script bbr

2022-12-17
 
1)- add exclucion de puertos udp en udpServer

2022-11-24

1)- corregido falla al eliminar host dns

2)- se agrego renovacion de usuario v2ray

2022-10-24

1)- fix code (limitador)

2)- fix code (visualizacion user ssh)

3)- admin de puertos activos movido a herramientas online
    
4)- nueva funcion administracion dns

2022-10-10

1)- generador de sub-dominios actualizado (online, y mas seguro)

3)- fix en opcion de actulizacion (muestra detalles de la actulizacion)

2)- fix instalador del script

2022-09-18

1)- fix checkuser

2022-09-09

1)- fix v2ray (instala v4.45.2 en lugar de v5.1.0)

2022-08-24

1)- slowdns ahora de ejecuta via servicios systemctl

2)- fix slowdns

3)- squid reconstruid

4)- fix, ubuntu 20,21,22

2022-08-24

1)- estilo instalado v2ray renovado

2)- mejorar en la instaalcion

3)- fix code v2ray

2022-08-22

1)- fix code

2)- se agrego CHEKUSER para apps que verificar el tiempo restante del usuario

2022-07-04

1)- fix code 

2022-07-01

1)- generar sub-dominios activa

2)- script generar dominios, nuevo codigo

3)- agrego genera domnio tipo A y NS en paralelo

## Historial de :point_right: [ACTUALIZACIONES](https://github.com/Farukbrowser/XCLANPro/blob/main/history.md)

#REPORTES DE BUSG POR:

[@Alexmod80](https://t.me/Alexmod80)

[@ander2314](https://t.me/ander2314)

[@arqimidez](https://t.me/arqimidez)

[@Walterlzz](https://t.me/Walterlzz)

[@Nelsongabriel32](https://t.me/Nelsongabriel32)

[@titoec593](https://t.me/titoec593)

[@Rg4bott](https://t.me/Rg4bott)

[@SonicBoom2021](https://t.me/SonicBoom2021)

[@CrakenYT](https://t.me/CrakenYT)

[@WillyAbel](https://t.me/WillyAbel)

[@JuanCdiaz99](https://t.me/JuanCdiaz99)

[@Trigger_Fulanito](https://t.me/Trigger_Fulanito)

[@Near365](https://t.me/Near365)

[@JManu287](https://t.me/JManu287)

[@Stain89mx](https://t.me/Stain89mx)

history.md
# Script XCLANPro :point_right: [instalacion y actualizaciones](https://github.com/Farukbrowser/XCLANPro/blob/main/README.md)

## Updates
2022-10-24

1)- fix code (limitador)

2)- fix code (visualizacion user ssh)

3)- admin de puertos activos movido a herramientas online
    
4)- nueva funcion administracion dns

2022-10-10

1)- generador de sub-dominios actualizado (online, y mas seguro)

3)- fix en opcion de actulizacion (muestra detalles de la actulizacion)

2)- fix instalador del script

2022-09-18

1)- fix checkuser

2022-09-09

1)- fix v2ray (instala v4.45.2 en lugar de v5.1.0)

2022-08-24

1)- slowdns ahora de ejecuta via servicios systemctl

2)- fix slowdns

3)- squid reconstruid

4)- fix, ubuntu 20,21,22

2022-08-24

1)- estilo instalado v2ray renovado

2)- mejorar en la instaalcion

3)- fix code v2ray

2022-08-22

1)- fix code

2)- se agrego CHEKUSER para apps que verificar el tiempo restante del usuario

2022-07-04

1)- fix code 

2022-07-01

1)- generar sub-dominios activa

2)- script generar dominios, nuevo codigo

3)- agrego genera domnio tipo A y NS en paralelo

2022-06-08

1)- agrego opcion de fix 'inciar con el sistema' DROPBEAR/SSL

2)- fix opcion para ([0] salir del vps). cierra la cesion

2022-06-01

1)- agrego explorador de archivos web (filebrowser)

2)- rescrivio code openvpn

3)- agrego opcion editar apariencia menu de protocolos

2022-05-26

1)- añadio opcion para ([0] salir del vps). cierra la cesion

2022-05-20

1)- fix menu clientes wireguard

2)- fix monitor de cuentas SSH/HWID/TOKEN

2022-05-19

1)- fix wireguard (para redes NAT)

2)- fix monitor de cuentas SSH

2022-05-17

1)- añadido WireGard install

2)- añadido administracion de cuentas wireguard

3)- modificacion de menu principal, (dinamico, acorde al protocolo activo)

4)- fix code

2022-05-05

1)- fix code

2)- se añadio modificacion de Reseller desde el botgen

2022-04-13

1)- fix bug limitador de expirados SSH/HWID/TOKEN

2022-04-12

1)- fix code

2022-04-11

1)- add menu manager ssl

2)- fix bug squid

3)- add menu manager squid

2022-03-31

1)- fix code

2022-03-21

1)- (V2RAY) se bloqueo la lectura de puertos UDP, para mejorar la visualizacion (el bug se esta en estudio.)

2022-03-20

1)- correcion de bug (cerrado de pueto dropbear) al aplicar/editar banner SSH

2022-03-16

1)- fix banner SSH/DROPBEAR

2)- se agrego un menu para edicion de banner SSH/DROPBEAR

2022-03-15

1)- Se agrego soporte para adminstracion de clinetes modos SSH/HWID/TOKEN

2)- Restaurar copis de usuarios online

3)- limitador de conecciones y expirados por separado

2022-03-11

1)- (fix bug) en inicio de limitador

2022-03-09

1)- fix bug backup de usuarios

2022-03-08

1)- fix visualizacion de actualizacion

2)- fix opcion reboot

3)- fix port node ws

# Script XCLANPro :point_right: [instalacion y actualizaciones](https://github.com/Farukbrowser/XCLANPro/blob/main/README.md)


online/clearLog.sh
#!/bin/bash

#source /etc/XCLANPro/module

opc=$1

clearlog(){
	filesBack=$(ls -lh /var/log/|awk '{print $9}')
	for backs in $filesBack; do
		rm -rf /var/log/$backs.*
	done
	files=$(ls -lh /var/log/|grep adm|grep "M\|G"|awk '{print $9}')
	for i in $files; do
		> /var/log/$i
	done

	echo "$(printf '%(%H:%M:%S)T')" > /root/clearlog.txt
}

autoClearLog(){
	title 'PROGRAMADOR DE TAREA CLEAR LOGS'
	if [[ ! -z $(crontab -l|grep 'clearLog.sh') ]]; then
	    msg -azu " Tarea programada cada $(msg -verd "[ $(crontab -l|grep 'clearLog.sh'|awk '{print $2}'|sed $'s/[^[:alnum:]\t]//g')HS ]")"
	    msg -bar
	    while :
	    do
		    in_opcion -nama 'Quitar tarea programada [S/N]'
		    del 2
		    case $opcion in
		      s|S)  crontab -l > /root/cron && sed -i '/clearLog.sh/ d' /root/cron && crontab /root/cron && rm /root/cron
					print_center -verd 'Tarea automatica removida!'
					enter
					return;;
		      n|N)  return;;
		      	*)  print_center -ama 'Selecciona S para si, N para no'
					sleep 2
					del 1
		    esac
	    done
	fi 

    in_opcion -nama 'PERIODO DE EJECUCION DE LA TAREA [1-12HS]'
    del 2
    if [[ $opcion =~ $numero ]]; then
		crontab -l > /root/cron
		echo "0 */$opcion * * * sudo /etc/XCLANPro/source/tool/clearLog.sh --clearlog" >> /root/cron
		crontab /root/cron
		rm /root/cron
		msg -azu " Tarea automatica programada cada: $(msg -verd "${opcion}HS")" && msg -bar && sleep 2
    else
		print_center -verm2 'ingresar solo numeros entre 1 y 12'
		sleep 2
		msg -bar
    fi
}

menuClearLog(){
	title 'HERRAMIENTA CLEAR LOGS'
	menu_func 'LIMPIAR LOGS' 'PROGRAMAR TAREA AUTOMATICA'
	back
	opcion=$(selection_fun 2)
	case $opcion in
		1)  print_center -ama 'limpiando logs'
			clearlog
			sleep 2
			del 1
			print_center -verd 'limpiesa de logs completa'
			enter;;
		2)autoClearLog;;
		0)return 1;;
	esac
}

if [[ $opc = '--clearlog' ]]; then
	clearlog
	exit
fi
while [[ $? -eq 0 ]]; do
	menuClearLog
done


Utils/badvpn/budp.sh
#!/bin/bash
#19/12/2019
clear
msg () {
local colors=""
if [[ ! -e $colors ]]; then
COLOR[0]='\033[1;37m' #BRAN='\033[1;37m'
COLOR[1]='\e[31m' #VERMELHO='\e[31m'
COLOR[2]='\e[32m' #VERDE='\e[32m'
COLOR[3]='\e[33m' #AMARELO='\e[33m'
COLOR[4]='\e[34m' #AZUL='\e[34m'
COLOR[5]='\e[91m' #MAGENTA='\e[35m'
COLOR[6]='\033[1;97m' #MAG='\033[1;36m'
else
local COL=0
for number in $(cat $colors); do
case $number in
1)COLOR[$COL]='\033[1;37m';;
2)COLOR[$COL]='\e[31m';;
3)COLOR[$COL]='\e[32m';;
4)COLOR[$COL]='\e[33m';;
5)COLOR[$COL]='\e[34m';;
6)COLOR[$COL]='\e[35m';;
7)COLOR[$COL]='\033[1;36m';;
esac
let COL++
done
fi
NEGRITO='\e[1m'
SEMCOR='\e[0m'
 case $1 in
  -ne)cor="${COLOR[1]}${NEGRITO}" && echo -ne "${cor}${2}${SEMCOR}";;
  -ama)cor="${COLOR[3]}${NEGRITO}" && echo -e "${cor}${2}${SEMCOR}";;
  -verm)cor="${COLOR[3]}${NEGRITO}[!] ${COLOR[1]}" && echo -e "${cor}${2}${SEMCOR}";;
  -verm2)cor="${COLOR[1]}${NEGRITO}" && echo -e "${cor}${2}${SEMCOR}";;
  -azu)cor="${COLOR[6]}${NEGRITO}" && echo -e "${cor}${2}${SEMCOR}";;
  -verd)cor="${COLOR[2]}${NEGRITO}" && echo -e "${cor}${2}${SEMCOR}";;
  -bra)cor="${COLOR[0]}${NEGRITO}" && echo -e "${cor}${2}${SEMCOR}";;
  "-bar2"|"-bar")cor="${COLOR[1]}=====================================================" && echo -e "${SEMCOR}${cor}${SEMCOR}";;
  -bar3)cor="${COLOR[1]}-----------------------------------------------------" && echo -e "${SEMCOR}${cor}${SEMCOR}";;
 esac
}

msg -bar
echo -e  "         INSTALADOR BADVPN-UDPGW @Rufu99" | lolcat
msg -bar

BadVPN () {
if [[ -z $(lsof -V -i tcp -P -n | grep -v "ESTABLISHED" |grep -v "COMMAND"|grep "badvpn-ud"|awk '{print $1}') ]]; then
    msg -ama "                  INICIADO BADVPN"
    msg -bar

echo -e "[Unit]
Description=BadVPN UDPGW Service
After=network.target\n
[Service]
Type=simple
User=root
WorkingDirectory=/root
ExecStart=/usr/bin/badvpn-udpgw --listen-addr 127.0.0.1:7300 --max-clients 1000 --max-connections-for-client 10
Restart=always
RestartSec=3s\n
[Install]
WantedBy=multi-user.target" > /etc/systemd/system/badvpn.service

    systemctl enable badvpn &>/dev/null
    systemctl start badvpn &>/dev/null
    sleep 2
    [[ -z $(lsof -V -i tcp -P -n | grep -v "ESTABLISHED" |grep -v "COMMAND"|grep "badvpn-ud"|awk '{print $1}') ]] && msg -verm2 "                  FALLA AL INICIAR" || msg -verd "                  BADVPN INICIADO" 
    msg -bar
    sleep 1
else
    msg -ama "                DETENIENDO BADVPN"
    msg -bar
    systemctl stop badvpn &>/dev/null
    systemctl disable badvpn &>/dev/null
    rm /etc/systemd/system/badvpn.service
    sleep 2
    [[ -z $(lsof -V -i tcp -P -n | grep -v "ESTABLISHED" |grep -v "COMMAND"|grep "badvpn-ud"|awk '{print $1}') ]] && msg -ama "                 BADVPN DETENIDO" || msg -verm2 "                FALLA AL DETENER"
    msg -bar
    sleep 1  
fi
unset st_badvpn
}

if [[ ! -e /root/udp-rufu ]]; then
	rm -rf /usr/bin/badvpn-udpgw &>/dev/null
	rm -rf /bin/badvpn-udpgw &>/dev/null
	touch /root/udp-rufu
fi
        if [[ ! -e /usr/bin/badvpn-udpgw ]]; then
            echo -ne "$(msg -azu " INSTALADO DEPENDECIAS...") "

            if apt install cmake -y &>/dev/null; then
                msg -verd "[OK]"
            else
                msg -verm2 "[fail]"
                slee 3
                return
            fi
            cd $HOME
            echo -ne "$(msg -azu " DESCARGANDO BADVPN......") "
            if wget https://github.com/Farukbrowser/XCLANPro/raw/main/Utils/badvpn/badvpn-master.zip &>/dev/null; then
                msg -verd "[OK]"
            else
                msg -verm2 "[fail]"
                slee 3
                return
            fi

            echo -ne "$(msg -azu " DESCOMPRIMIENDO.........") "
            if unzip badvpn-master.zip &>/dev/null; then
                msg -verd "[OK]"
            else
                msg -verm2 "[fail]"
                slee 3
                return
            fi
            cd badvpn-master
            mkdir build
            cd build

            echo -ne "$(msg -azu " COMPILANDO BADVPN.......") "
            if cmake .. -DCMAKE_INSTALL_PREFIX="/" -DBUILD_NOTHING_BY_DEFAULT=1 -DBUILD_UDPGW=1 &>/dev/null && make install &>/dev/null; then
                msg -verd "[OK]"
            else
                msg -verm2 "[fail]"
                slee 3
                return
            fi
            cd $HOME
            rm badvpn-master.zip &>/dev/null
        fi
BadVPN


online/firewall-VPS.sh
#!/bin/bash
clear
#source /etc/XCLANPro/module

smtp_port="25,26,465,587"
pop3_port="109,110,995"
imap_port="143,218,220,993"
other_port="24,50,57,105,106,158,209,1109,24554,60177,60179"

firewall_port="$(pwd)/firewall_port" && [[ ! -e ${firewall_port} ]] && echo -e "$smtp_port\n$pop3_port\n$imap_port\n$other_port" > ${firewall_port}

bt_key_word="torrent
.torrent
peer_id=
announce
info_hash
get_peers
find_node
BitTorrent
announce_peer
BitTorrent protocol
announce.php?passkey=
magnet:
xunlei
sandai
Thunder
XLLiveUD"

firewall_txt="$(pwd)/firewall_txt" && [[ ! -e ${firewall_txt} ]] && echo -e "${bt_key_word}" > ${firewall_txt}

check_sys(){
	if [[ -f /etc/redhat-release ]]; then
		release="centos"
	elif cat /etc/issue | grep -q -E -i "debian"; then
		release="debian"
	elif cat /etc/issue | grep -q -E -i "ubuntu"; then
		release="ubuntu"
	elif cat /etc/issue | grep -q -E -i "centos|red hat|redhat"; then
		release="centos"
	elif cat /proc/version | grep -q -E -i "debian"; then
		release="debian"
	elif cat /proc/version | grep -q -E -i "ubuntu"; then
		release="ubuntu"
	elif cat /proc/version | grep -q -E -i "centos|red hat|redhat"; then
		release="centos"
    fi
	bit=`uname -m`
}
check_BT(){
	Cat_KEY_WORDS
	BT_KEY_WORDS=$(echo -e "$Ban_KEY_WORDS_list"|grep "torrent")
}
check_SPAM(){
	Cat_PORT
	SPAM_PORT=$(echo -e "$Ban_PORT_list"|grep "${smtp_port}")
}
Cat_PORT(){
	Ban_PORT_list=$(iptables -t filter -L OUTPUT -nvx --line-numbers|grep "REJECT"|awk '{print $13}')
}
Cat_KEY_WORDS(){
	Ban_KEY_WORDS_list=""
	Ban_KEY_WORDS_v6_list=""
	if [[ ! -z ${v6iptables} ]]; then
		Ban_KEY_WORDS_v6_text=$(${v6iptables} -t mangle -L OUTPUT -nvx --line-numbers|grep "DROP")
		Ban_KEY_WORDS_v6_list=$(echo -e "${Ban_KEY_WORDS_v6_text}"|sed -r 's/.*\"(.+)\".*/\1/')
	fi
	Ban_KEY_WORDS_text=$(${v4iptables} -t mangle -L OUTPUT -nvx --line-numbers|grep "DROP")
	Ban_KEY_WORDS_list=$(echo -e "${Ban_KEY_WORDS_text}"|sed -r 's/.*\"(.+)\".*/\1/')
}
View_PORT(){
	Cat_PORT
	if [[ ! -z ${Ban_PORT_list} ]]; then
		port_list=${Ban_PORT_list}
		print_center -azu "LISTA DE PUERTO BLOQUEADOS \e[32m[ON]"
		msg -bar
		echo -e "\033[1;97m${port_list}"
	else
		port_list=$(cat $firewall_port)
		print_center -azu "LISTA DE PUERTO BLOQUEADOS \e[31m[OFF]"
		msg -bar
		echo -e "\033[1;97m${port_list}"
	fi
}
View_KEY_WORDS(){
	Cat_KEY_WORDS
	if [[ ! "${op_1[1]}" = '1' ]]; then
		txt_list=${Ban_KEY_WORDS_list}
		print_center -azu "lista de palabras claves \e[32m[ON]"
		msg -bar
		msg -azu "${txt_list}"
	else
		txt_list=$(cat ${firewall_txt})
		print_center -azu "lista de palabras claves \e[31m[OFF]"
		msg -bar
		msg -azu "${txt_list}"
	fi
}
View_KEY_WORDS_2(){
	Cat_KEY_WORDS
	if [[ "${op_1[1]}" = '1' ]]; then
		txt_list=${Ban_KEY_WORDS_list}
		print_center -azu "lista de palabras claves \e[32m[ON]"
		msg -bar
		msg -azu "${txt_list}"
	else
		txt_list=$(cat ${firewall_txt})
		print_center -azu "lista de palabras claves \e[31m[OFF]"
		msg -bar
		msg -azu "${txt_list}"
	fi
}
View_ALL(){
	clear
	msg -bar
	View_PORT
	msg -bar
	View_KEY_WORDS_2
}
Save_iptables_v4_v6(){
	if [[ ${release} == "centos" ]]; then
		if [[ ! -z "$v6iptables" ]]; then
			service ip6tables save
			chkconfig --level 2345 ip6tables on
		fi
		service iptables save
		chkconfig --level 2345 iptables on
	else
		if [[ ! -z "$v6iptables" ]]; then
			ip6tables-save > /etc/ip6tables.up.rules
			echo -e "#!/bin/bash\n/sbin/iptables-restore < /etc/iptables.up.rules\n/sbin/ip6tables-restore < /etc/ip6tables.up.rules" > /etc/network/if-pre-up.d/iptables
		else
			echo -e "#!/bin/bash\n/sbin/iptables-restore < /etc/iptables.up.rules" > /etc/network/if-pre-up.d/iptables
		fi
		iptables-save > /etc/iptables.up.rules
		chmod +x /etc/network/if-pre-up.d/iptables
	fi
}
Set_key_word() { $1 -t mangle -$3 OUTPUT -m string --string "$2" --algo bm --to 65535 -j DROP; }
Set_tcp_port() {
	[[ "$1" = "$v4iptables" ]] && $1 -t filter -$3 OUTPUT -p tcp -m multiport --dports "$2" -m state --state NEW,ESTABLISHED -j REJECT --reject-with icmp-port-unreachable
	[[ "$1" = "$v6iptables" ]] && $1 -t filter -$3 OUTPUT -p tcp -m multiport --dports "$2" -m state --state NEW,ESTABLISHED -j REJECT --reject-with tcp-reset
}
Set_udp_port() { $1 -t filter -$3 OUTPUT -p udp -m multiport --dports "$2" -j DROP; }


Set_SPAM_Code_v4(){
	while read line; do
		Set_tcp_port $v4iptables "$line" $s
		Set_udp_port $v4iptables "$line" $s
	done <<< $(cat "$firewall_port")
}

Set_SPAM_Code_v4_back(){
	for i in ${smtp_port} ${pop3_port} ${imap_port} ${other_port}
		do
		Set_tcp_port $v4iptables "$i" $s
		Set_udp_port $v4iptables "$i" $s
	done
}

Set_SPAM_Code_v4_v6(){
	while read line; do
		for j in $v4iptables $v6iptables
		do
			Set_tcp_port $j "$line" $s
			Set_udp_port $j "$line" $s
		done
	done <<< $(cat "$firewall_port")
}

Set_SPAM_Code_v4_v6_back(){
	for i in ${smtp_port} ${pop3_port} ${imap_port} ${other_port}
	do
		for j in $v4iptables $v6iptables
		do
			Set_tcp_port $j "$i" $s
			Set_udp_port $j "$i" $s
		done
	done
}
Set_PORT(){
	if [[ -n "$v4iptables" ]] && [[ -n "$v6iptables" ]]; then
		Set_tcp_port $v4iptables $PORT $s
		Set_udp_port $v4iptables $PORT $s
		Set_tcp_port $v6iptables $PORT $s
		Set_udp_port $v6iptables $PORT $s
	elif [[ -n "$v4iptables" ]]; then
		Set_tcp_port $v4iptables $PORT $s
		Set_udp_port $v4iptables $PORT $s
	fi
	Save_iptables_v4_v6
}
Set_KEY_WORDS(){
	key_word_num=$(echo -e "${key_word}"|wc -l)
	for((integer = 1; integer <= ${key_word_num}; integer++))
		do
			i=$(echo -e "${key_word}"|sed -n "${integer}p")
			Set_key_word $v4iptables "$i" $s
			[[ ! -z "$v6iptables" ]] && Set_key_word $v6iptables "$i" $s
	done
	Save_iptables_v4_v6
}
Set_BT(){
	key_word=$(cat ${firewall_txt})
	Set_KEY_WORDS
	Save_iptables_v4_v6
}
Set_SPAM(){
	if [[ -n "$v4iptables" ]] && [[ -n "$v6iptables" ]]; then
		Set_SPAM_Code_v4_v6
	elif [[ -n "$v4iptables" ]]; then
		Set_SPAM_Code_v4
	fi
	Save_iptables_v4_v6
}
Set_ALL(){
	Set_BT
	Set_SPAM
}
Ban_BT(){
	check_BT
	s="A"
	Set_BT
	clear
	msg -bar
	View_KEY_WORDS
	msg -bar
	print_center -verd "Torrent bloqueados y Palabras Claves!"
	enter
	return
}
Ban_SPAM(){
	check_SPAM
	s="A"
	Set_SPAM
	clear
	msg -bar
	View_PORT
	msg -bar
	print_center -verd "Puertos SPAM Bloqueados !"
	enter
	return
}
UnBan_BT(){
	check_BT
	s="D"
	Set_BT
	clear
	msg -bar
	View_KEY_WORDS
	msg -bar
	print_center -verd "Torrent Desbloqueados y Palabras Claves!"
	enter
	return
}
UnBan_SPAM(){
	check_SPAM
	s="D"
	Set_SPAM
	clear
	msg -bar
	View_PORT
	msg -bar
	print_center -verd "Puertos de SPAM Desbloqueados !"
	enter
	return
}

ENTER_Ban_KEY_WORDS_type(){
	Cat_KEY_WORDS
	title -ama "AGREGAR/QUITAR PALABRAS/URL/ARCHIVO"

	if [[ "${op_1[1]}" = '1' ]]; then
		txt_list=${Ban_KEY_WORDS_list}
		print_center -azu "lista de palabras claves \e[32m[ON]"
		msg -bar3
		msg -azu "${txt_list}"
	else
		txt_list=$(cat ${firewall_txt})
		print_center -azu "lista de palabras claves \e[31m[OFF]"
		msg -bar3
		msg -azu "${txt_list}"
	fi
	msg -bar3
	msg -ama "Archivo local: --file /direcion/del/archivo.txt"
	msg -ama "Archivo online: --url http://direcion.del:archivo.txt"
	msg -bar3
	#echo -e " $(msg -verm3 "╭╼╼╼╼╼╼╼╼╼╼[")$(msg -azu "ingresa Palabra/url/archivo")$(msg -verm3 "]")"
	#echo -ne " $(msg -verm3 "╰╼")\033[37;1m> "
	#read option

	in_opcion_down 'ingresa Palabra/url/archivo'
	option=$opcion

	[[ -z $option ]] && return 1

	opt=$(echo "$option"|awk -F ' ' '{print $1}')

	case $opt in
		-u|--url) opn='' && opc=$(echo "$option"|awk -F ' ' '{print $2}') && link=$(wget --no-check-certificate -t3 -T5 -qO- "${opc}")
				    for i in `echo "${link}"`; do
				   		[[ $(echo "$txt_list"|grep -w "$i") ]] && continue
				   		opn+="$i\n"
				   	done
				   	echo -e "${opn}" >> firewall_txt
				   	sed -i '/^$/d' ${firewall_txt}
				   	key_word="$(echo -e "${opn}")";;
		-f|--file) opn='' && opc=$(echo "$option"|awk -F ' ' '{print $2}')
				    for i in `cat "${opc}"`; do
				   		[[ $(echo "$txt_list"|grep -w "$i") ]] && continue
				   		opn+="$i\n"
				   	done
				   	echo -e "${opn}" >> firewall_txt
				   	sed -i '/^$/d' ${firewall_txt}
				   	key_word="$(echo -e "${opn}")";;
		*)  if [[ $(echo "$txt_list"|grep -w "$option") ]]; then
				s="D"
				sed -i "/$option/d" ${firewall_txt}
			else
				echo "$option" >> firewall_txt
			fi
			sed -i '/^$/d' ${firewall_txt}
			key_word=$option;;
	esac
}

ENTER_Ban_PORT(){
	title -ama "AGREGAR/QUITAR PUERTOS"
	Cat_PORT
	if [[ ! -z ${Ban_PORT_list} ]]; then
		port_list=${Ban_PORT_list}
		print_center -azu "LISTA DE PUERTO BLOQUEADOS \e[32m[ON]"
		msg -bar3
		echo -e "\033[1;97m${port_list}"
		msg -bar3
	else
		port_list=$(cat $firewall_port)
		print_center -azu "LISTA DE PUERTO BLOQUEADOS \e[31m[OFF]"
		msg -bar3
		echo -e "\033[1;97m${port_list}"
		msg -bar3
	fi
	if [[ ${Ban_PORT_Type_1} != "1" ]]; then
		print_center -ama "modo de ingreso de puertos"
		msg -bar3
		msg -ama " Puerto único:     \e[32m25"
		msg -ama " Multiples pertos: \e[32m25,26,465,587"
		msg -ama " Rango de puertos: \e[32m25:587"
		msg -bar3
	fi
	
	PORT=$(in_opcion "Intro se cancela por defecto")
	[[ -z "${PORT}" ]] && return 1

	if [[ $(echo "${PORT:0:1}"|grep '0\|[A-Za-z]\|,\|:\|\$\|#\|\[\|\]\|/\|?\|@\|!\|&\|(\|)\|*\|+\|;\|=\|%\|"') ]]; then
		del 1
		print_center -verm2 'formato invalido!'
		sleep 2
		return 1
	elif [[ $(echo "${PORT}"|grep '[A-Za-z]\|\$\|#\|\[\|\]\|/\|?\|@\|!\|&\|(\|)\|*\|+\|;\|=\|%\|"') ]]; then
		del 1
		print_center -verm2 'formato invalido!'
		sleep 2
		return 1
	fi

	if [[ $(echo "${port_list}"|grep -w "$PORT") ]]; then
		s="D"
		delP="$PORT"
		delR=$(echo "${port_list}"|grep -w "$delP")


		if [[ $(echo "$delR"|grep ':') ]]; then
			sed -i "/:$delP/d" ${firewall_port}
			sed -i "/$delP:/d" ${firewall_port}
			PORT="$delR"
		elif [[ $(echo "$delR"|grep ',') ]]; then
			sed -i "s/,$delP//g" ${firewall_port}
			sed -i "s/$delP,//g" ${firewall_port}

			PORT="$delR"
			[[ "${op_2[1]}" = '1' ]] && Set_PORT

			PORT=$(echo "$delR"|sed "/,$delP/d"|sed "/$delP,/d")
			s="A"
		else
			sed -i "/$delP/d" ${firewall_port}
			PORT="$delR"
		fi
	else
		s="A"
		sed -i '/^$/d' ${firewall_port}
		echo "$PORT" >> ${firewall_port}
	fi
	return 0
}

Ban_PORT(){
	s="A"
	while true; do
		ENTER_Ban_PORT
		[[ $? -eq 1 ]] && break
		[[ "${op_2[1]}" = '1' ]] && [[ -z $PORT ]] && Set_PORT
	done
}

Ban_KEY_WORDS(){
	s="A"
	while true; do
		ENTER_Ban_KEY_WORDS_type
		[[ $? -eq 1 ]] && break
		[[ "${op_1[1]}" = '1' ]] && Set_KEY_WORDS
	done
}

check_iptables(){
	v4iptables=`iptables -V`
	v6iptables=`ip6tables -V`
	if [[ ! -z ${v4iptables} ]]; then
		v4iptables="iptables"
		if [[ ! -z ${v6iptables} ]]; then
			v6iptables="ip6tables"
		fi
	else
		print_center -ama "El firewall de iptables no está instalado!\nPor favor, instale el firewall de iptables：\nCentOS Sistema： yum install iptables -y\nDebian/Ubuntu Sistema： apt-get install iptables -y"
	fi
}

menu_fw(){
	check_sys
	check_iptables

	unset opcion

	check_BT; op_1=($([[ -z ${BT_KEY_WORDS} ]] && msg -verm2 "[OFF]" || msg -verd "[ON]") $([[ -z ${BT_KEY_WORDS} ]] && echo '0' || echo '1'))
	check_SPAM; op_2=($([[ -z ${SPAM_PORT} ]] && msg -verm2 "[OFF]" || msg -verd "[ON]") $([[ -z ${SPAM_PORT} ]] && echo '0' || echo '1'))

	title -ama "Panel Firewall By @Rufu99"
 menu_func "Bloq/Desbl Torrent, Palabras Clave ${op_1}" \
 "-bar Bloq/Desbl Puertos SPAM ${op_2}" \
 "Bloq/Desbl Puerto SPAM custom" \
 "-bar Bloq/Desbl Palabras Clave custom" \
 "Ver lista de reglas firewall"

 back

 	opcion=$(selection_fun 5)
 	case ${opcion} in
 		1)if [[ "${op_1[1]}" = '1' ]]; then UnBan_BT; else Ban_BT; fi;;
 		2)if [[ "${op_2[1]}" = '1' ]]; then UnBan_SPAM; else Ban_SPAM; fi;;
 		3)Ban_PORT;;
 		4)Ban_KEY_WORDS;;
 		5)View_ALL && enter;;
 		0)return 1;;
 	esac
}

while [[ $? -eq 0 ]]; do
	menu_fw
done
